include/nbl/builtin/glsl/shapes/rectangle.glsl

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_GLSL_SHAPES_RECTANGLE_INCLUDED_
+#define _NBL_BUILTIN_GLSL_SHAPES_RECTANGLE_INCLUDED_
+
+#include <nbl/builtin/glsl/math/functions.glsl>
+
+struct SphRect {
+    mat3 basis;
+    vec3 r0;
+    vec3 r1;
+};
```

Comment:
this should store only the bare essentials you need for computing Solid Angle and/or sampling

Reply:
`basis` doesn't really need to be kept around

Reply:
you only need
```glsl
struct nbl_glsl_SphericalRectangle
{
   vec3 origin;
   vec2 extents;
};
```
I think

Reply:
I eliminated basis from the struct, but I think we still need to store r0. Especially since if we would need two separate functions to switch between projected solid angle sampling. Computing r0 again isn't really heavy tho. (just on matrix multiplication)

Reply:
From api perspective,
doing this:
```cpp
struct SphRect {
  vec3 offset;
  mat3 basis;
  vec2 extent;
}
struct SphRectCoords {
  vec3 r0;
  vec3 r1;
  uint zSignFlipMask;
}
```
and have two functions like
SphRect SphericalRectangle_SphericalRectangle(vec3 offset, mat3 basis, vec2 extent);
SphRectCoords SphericalRectangle_getCoords(vec3 oberserver);
might be better.

How do you think of this?

Reply:
Always think about the long-game, which is sampling a Spherical OBB.

In which case you'll want to have an API which accepts precomputed `r0` and `extent`.

You can compute `r1` on  the fly
https://github.com/Devsh-Graphics-Programming/Nabla/pull/340#discussion_r840025227

So all you really need is `r0` and `extents` to perform Solid Angle computation or sampling.

At the end you can have an extra function which will turn your local uv rectangle coordinates into a point in world-space given a rectangle mat2x3 basis for UV + position/origin.

Reply:
Done. The API now looks solid to me.